File: src/Scanner.Core/Checks/MissingLabelCheck.cs
using System.Text.RegularExpressions;
using Scanner.Core.Rules;

namespace Scanner.Core.Checks;

/// <summary>
/// Checks form controls for missing accessible labels.
/// </summary>
public sealed class MissingLabelCheck : ICheck
{
    private static readonly Regex InputRegex = new("<(input|select|textarea)(?<attrs>[^>]*)>", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex LabelRegex = new("<label[^>]*for=\"(?<id>[^\"]+)\"[^>]*>", RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <inheritdoc />
    public string Id => "missing-label";

    /// <inheritdoc />
    public IReadOnlyCollection<string> ApplicableKinds { get; } = new[] { "html", "htm", "cshtml", "razor" };

    /// <inheritdoc />
    public IEnumerable<Issue> Run(CheckContext context, RuleDefinition rule)
    {
        var issues = new List<Issue>();
        var labels = LabelRegex.Matches(context.Content).Select(match => match.Groups["id"].Value).ToHashSet(StringComparer.OrdinalIgnoreCase);

        foreach (Match match in InputRegex.Matches(context.Content))
        {
            var attrs = match.Groups["attrs"].Value;
            var type = AttributeParser.GetAttributeValue(attrs, "type");
            if (string.Equals(type, "hidden", StringComparison.OrdinalIgnoreCase))
            {
                continue;
            }

            var ariaLabel = AttributeParser.GetAttributeValue(attrs, "aria-label");
            if (!string.IsNullOrWhiteSpace(ariaLabel))
            {
                continue;
            }

            var ariaLabelledBy = AttributeParser.GetAttributeValue(attrs, "aria-labelledby");
            if (!string.IsNullOrWhiteSpace(ariaLabelledBy))
            {
                continue;
            }

            var id = AttributeParser.GetAttributeValue(attrs, "id");
            if (!string.IsNullOrWhiteSpace(id) && labels.Contains(id))
            {
                continue;
            }

            var line = TextUtilities.GetLineNumber(context.Content, match.Index);
            issues.Add(new Issue(rule.Id, Id, context.FilePath, line, "Form control missing accessible label.", match.Value));
        }

        return issues;
    }
}

File: node_engine/src/checks/MissingLabelCheck.js
const { getAttributeValue } = require("./AttributeParser");
const {
  collectLabelForIds,
  collectLabelForIdsFromDocument,
  collectLabelRanges,
  collectElementIds,
  collectElementIdsFromDocument,
  hasAriaLabel,
  hasValidAriaLabelledBy,
  hasLabelForId,
  isWithinLabel
} = require("./AccessibleNameUtilities");
const { getLineNumber } = require("./TextUtilities");

const inputRegex = /<(input|select|textarea)(?<attrs>[^>]*)>/gi;

const MissingLabelCheck = {
  id: "missing-label",
  applicableKinds: ["html", "htm", "cshtml", "razor"],
  run(context, rule) {
    const issues = [];
    if (context.document?.querySelectorAll) {
      const labelForIds = collectLabelForIdsFromDocument(context.document);
      const elementIds = collectElementIdsFromDocument(context.document);
      const elementPositions = Array.from(context.content.matchAll(/<(input|select|textarea)\b/gi))
        .map((match) => match.index);
      let positionIndex = 0;

      for (const control of context.document.querySelectorAll("input, select, textarea")) {
        const matchIndex = elementPositions[positionIndex] ?? 0;
        positionIndex += 1;
        const type = control.getAttribute("type");
        if (type && type.toLowerCase() === "hidden") {
          continue;
        }

        const ariaLabel = control.getAttribute("aria-label");
        if (hasAriaLabel(ariaLabel)) {
          continue;
        }

        const labelledBy = control.getAttribute("aria-labelledby");
        if (hasValidAriaLabelledBy(labelledBy, elementIds)) {
          continue;
        }

        const id = control.getAttribute("id");
        if (hasLabelForId(id, labelForIds)) {
          continue;
        }

        if (control.closest("label")) {
          continue;
        }

        const evidence = control.outerHTML;
        issues.push({
          ruleId: rule.id,
          checkId: MissingLabelCheck.id,
          filePath: context.filePath,
          line: getLineNumber(context.content, matchIndex),
          message: "Form control missing accessible label.",
          evidence
        });
      }

      return issues;
    }

    const labels = collectLabelForIds(context.content);
    const labelRanges = collectLabelRanges(context.content);
    const elementIds = collectElementIds(context.content);

    for (const match of context.content.matchAll(inputRegex)) {
      const attrs = match.groups.attrs;
      const type = getAttributeValue(attrs, "type");
      if (type && type.toLowerCase() === "hidden") {
        continue;
      }

      const ariaLabel = getAttributeValue(attrs, "aria-label");
      if (hasAriaLabel(ariaLabel)) {
        continue;
      }

      const ariaLabelledBy = getAttributeValue(attrs, "aria-labelledby");
      if (hasValidAriaLabelledBy(ariaLabelledBy, elementIds)) {
        continue;
      }

      const id = getAttributeValue(attrs, "id");
      if (hasLabelForId(id, labels)) {
        continue;
      }

      if (isWithinLabel(match.index, labelRanges)) {
        continue;
      }

      const line = getLineNumber(context.content, match.index);
      issues.push({
        ruleId: rule.id,
        checkId: MissingLabelCheck.id,
        filePath: context.filePath,
        line,
        message: "Form control missing accessible label.",
        evidence: match[0]
      });
    }

    return issues;
  }
};

module.exports = { MissingLabelCheck };
